### ISSUE TITLE
Incorrect extend errors since v2.4 update

### ISSUE THREAD

**Issue by samchon** [Microsoft/TypeScript#17123]
I've been publishing a library compiled by the `TypeScript`. Since the `v2.4` update, my library fails to compile throwing such error:
  - https://github.com/samchon/tstl/issues/5

```bash
error TS2415: Class 'MapContainer<Key, T>' incorrectly extends base class 'Container<Pair<Key, T>>'.
error TS2415: Class 'MapIterator<Key, T>' incorrectly extends base class '_ListIteratorBase<Pair<Key, T>>'.
error TS2344: Type 'SetIterator<T>' does not satisfy the constraint '_ListIteratorBase<T>'.
...
...
```

It was not a problem to extending a class and overriding some methods to return (or get parameter) sub-type of origin. Thus, below example code, there hasn't been any error until the `v2.3`. However, since the `v2.4` update, below code throws `incorretly extends base class` error.

I want to know that such errors, since `v2.4`, are by whether:
  1. intended error of TypeScript-compiler; stricter type checking
  2. domain error of mis-update.

```typescript
namespace std.base
{
    //------------------------------------------------
    // ABSTRACT CLASSES
    //------------------------------------------------
    export abstract class Container<T>
    {
        public abstract begin(): Iterator<T>;
        public abstract end(): Iterator<T>;
        public erase(first: Iterator<T>, last: Iterator<T>): Iterator<T>;
    }
    export abstract class Iterator<T>
    {
        public abstract source(): Container<T>;
        public abstract prev(): Iterator<T>;
        public abstract next(): Iterator<T>;
    }
}

namespace std
{
    //------------------------------------------------
    // DERIVED CLASSES
    //------------------------------------------------
    // DERIVED CONTAINERS
    export class Vector<T> extends base.Container<T>
    {
        public begin(): VectorIterator<T>;
        public end(): VectorIterator<T>;
        public erase(first: VectorIterator<T>, last: VectorIterator<T>): VectorIterator<T>;
    }
    export class List<T> extends base.Container<T>
    {
        public begin(): ListIterator<T>;
        public end(): ListIterator<T>;
        public erase(first: ListIterator<T>, last: ListIterator<T>): ListIterator<T>;
    }

    // DERIVED ITERATORS
    export class VectorIterator<T> extends base.Iterator<T>
    {
        public source(): Vector<T>;
        public prev(): VectorIterator<T>;
        public next(): VectorIterator<T>;
    }
    export class ListIterator<T> extends base.Iterator<T>
    {
        public source(): List<T>;
        public prev(): ListIterator<T>;
        public next(): ListIterator<T>;
    }
}
```

**Comment by andy-ms** [comment]
Noticed this:
```ts

abstract class Container<T> {
    erase<U extends T>(begin: Iterator<U>, end: Iterator<U>): Iterator<T>;
}

abstract class SetContainer<T> extends Container<T> {
    erase(begin: SetIterator<T>, end: SetIterator<T>): SetIterator<T>;
}
```

You need `erase<U extends T>` in the subclass to have a compatible signature. This is a new error due to #16368. Previously we would squash all generics to `any` before comparing signatures.

**Comment by MatejQ** [comment]
I am getting this "incorrectly extends..." error too. However, my classes are not that simple and I cannot seem to be able to identify what exactly is wrong. It compiled fine in TypeScript 2.3. Is it possible to include more specific reason why TypeScript thinks it incorrectly extends parent class?

**Comment by andy-ms** [comment]
@MatejQ Usually TypeScript will drill down into the method that extends the base method incorrectly; maybe the tool you're using to display errors is hiding this? Try running `tsc` from the command line?

**Comment by MatejQ** [comment]
@andy-ms thanks, that really helped. I had to set the output of MSBuild in visual studio to detailed and then I got all the specific messages from tsc. I hope this helps someone else too.

**Comment by mhegazy** [comment]
based on https://github.com/Microsoft/TypeScript/issues/17123#issuecomment-314817188, this seems like the expected behavior.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
